Enhance LAL to support convert LogData to DatabaseSlowStatement. (#9575)
apache/skywalking: docs/en/concepts-and-designs/lal.md

## Filter

> A filter is a group of [parser](#parser), [extractor](#extractor), [SlowSql](#slowsql) and [sink](#sink). Users can use one or more filters

Reviewer: I'd expect `extractor` to includes all operations that extract other things than log itself, can you move the `slowSql` into `extractor`, like `metrics` extractor? By doing this, we can reuse the logics like, service name, instance name, timestamp, layer, etc.

Author: OK, I will change it soon.

Note that when you put `regexp` in an `if` statement, you need to surround the expression with `()`
like `regexp(<the expression>)`, instead of `regexp <the expression>`.

- `tag`

`tag` function provide a convenient way to get the data of tags.

We can add tags like following:
``` JSON
[
   {
      "tags":{
         "data":[
            {
               "key":"TEST_KEY",
               "value":"TEST_VALUE"
            }
         ]
      },
      "body":{
         ...
      }
      ...
   }
]
``` 

> And we can use this method to get the data of tags.

Reviewer: ```suggestion
And we can use this method to get the value of the tag key `TEST_KEY`.
```

    exp: http_response_time.sum(['le', 'service', 'instance']).increase('PT5M').histogram().histogram_percentile([50,70,90,99])
```

> ### SlowSql

Reviewer 2: Is this original log to be sinked or abandoned? Or does it depend on further LAL? If so, I think we should abandon this logs from persistence in the example with comments to explain about why.

> SlowSql aim to convert LogData to DatabaseSlowStatement. It extract data from `parsed` result and save them as DatabaseSlowStatement.

Reviewer 2: ```suggestion
SlowSql aims to convert LogData to DatabaseSlowStatement. It extracts data from `parsed` result and saves them as DatabaseSlowStatement for OAL engine.
```

    exp: http_response_time.sum(['le', 'service', 'instance']).increase('PT5M').histogram().histogram_percentile([50,70,90,99])
```

### SlowSql

SlowSql aims to convert LogData to DatabaseSlowStatement. It extracts data from `parsed` result and save them as DatabaseSlowStatement. SlowSql will not abort or edit logs, you can use other LAL for further processing.
We need to set a log tag("isSlowSql") to enable it. If logs sent to OAP does not have this tag, slowSql processing will be skipped.The example to set the tag is following:
``` json

> [{"tags":{"data":[{"key":"isSlowSql","value":"true"}]},"body":{"json":{"json":"{\"time\":\"20220906153959\",\"id\":\"cb92c1a5b-2691e-fb2f-457a-9c72a392d9ed\",\"service\":\"root[root]@[localhost]\",\"statement\":\"select sleep(2);\",\"layer\":\"MYSQL\",\"query_time\":2000}"}},"service":"root[root]@[localhost]"}]

Reviewer 2: Could you format this JSON?

persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `latency`

`latency` extracts the latency from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be
persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `id`

`id` extracts the id from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be persisted (if not
dropped) and is used to associate with TopNDatabaseStatement.

- `layer`

`layer` extracts the [layer](../../../oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/Layer.java) from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be persisted (if
not dropped) and is used to associate with service.

Example :

```groovy
filter {
        json{
        }
        slowSql {
          serviceName parsed.service as String
          id parsed.id as String
          statement parsed.statement as String
          latency parsed.query_time as Long
          layer parsed.layer as String
          timeBucket parsed.time as Long
        }

>       }

Reviewer 2: ```suggestion
      }
}
```

Author: I think an extra “}” is not needed.

Reviewer 2: <img width="545" alt="image" src="https://user-images.githubusercontent.com/5441976/188909382-51f3235e-e8ee-4298-a3dd-a06ea8793963.png">

I noticed this is not well formated.

Author: It is my fault , is this OK?
![image](https://user-images.githubusercontent.com/74546965/188910278-682084d9-9099-4e48-addb-4bc2fa751013.png)

Reviewer 2: This should be better.

    exp: http_response_time.sum(['le', 'service', 'instance']).increase('PT5M').histogram().histogram_percentile([50,70,90,99])
```

### SlowSql

SlowSql aims to convert LogData to DatabaseSlowStatement. It extracts data from `parsed` result and save them as DatabaseSlowStatement. SlowSql will not abort or edit logs, you can use other LAL for further processing.
We need to set a log tag("isSlowSql") to enable it. If logs sent to OAP does not have this tag, slowSql processing will be skipped.The example to set the tag is following:
``` json
[
   {
      "tags":{
         "data":[
            {
               "key":"isSlowSql",

>                "value":"true"

Reviewer 2: I am thinking about the naming of this tag. As an official tag, we should define it as

`key`=`LOG_KIND` `value`=`SLOW_SQL`.

Could you adjust codes and docs to this?

Author: I get it.

- `timeBucket`

`timeBucket` extracts the time bucket from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be
persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `statement`

`statement` extracts the time bucket from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be
persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `latency`

`latency` extracts the latency from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be
persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `id`

`id` extracts the id from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be persisted (if not
dropped) and is used to associate with TopNDatabaseStatement.

- `layer`

`layer` extracts the [layer](../../../oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/Layer.java) from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be persisted (if
not dropped) and is used to associate with service.

Example :

```groovy
filter {
        json{
        }

>         if (log.tags.getDataList().find{tag1->tag1.getKey() == "isSlowSql"}.getValue() == "true") {

Reviewer: This can be a trouble for the users who don't understand the Groovy syntax, let's simplify this, can you add a convenient getter method `String tag(String key)` in `ExtractorSpec` class, and change this example to:

```groovy
filter {
  json {
  }
  if (log.tag("isSlowSql") == "true") {
  }
}
```

Author: I understand.

Author: @kezhenxu94 Modifying Class `ExtractorSpec` may not have this effect, maybe I need to modify the 
Class `LALPrecompiledExtension`?

Reviewer: > @kezhenxu94 Modifying Class `ExtractorSpec` may not have this effect, maybe I need to modify the
> Class `LALPrecompiledExtension` and `AbstractSpec`?

Hi, feel free to modify those if needed, I just provided an idea and it's not complete.

Author: I have added a global method to resolve it.
```
filter {
        json{
        }
        extractor{
          layer parsed.layer as String
          service parsed.service as String
          timestamp parsed.time as String
          if (tag("LOG_KIND") == "SLOW_SQL") {
             slowSql {
                      id parsed.id as String
                      statement parsed.statement as String
                      latency parsed.query_time as Long
                     }
          }
        }
      }
```

    exp: http_response_time.sum(['le', 'service', 'instance']).increase('PT5M').histogram().histogram_percentile([50,70,90,99])
```

> #### SlowSql

Reviewer 2: ```suggestion
- SlowSql
```

I think `metrics` uses this format?

Author: But method in SlowSQL use this format too . it will can not distinguish.

Reviewer 2: You could read the preview page of markdown, https://github.com/apache/skywalking/blob/8d4bafc92fd8adda5c61eaf0a1970b127c0693ca/docs/en/concepts-and-designs/lal.md

<img width="279" alt="image" src="https://user-images.githubusercontent.com/5441976/189074830-18947c4e-3882-430e-9131-c93a3f48ea93.png">

This is totally different from others.

SlowSql aims to convert LogData to DatabaseSlowStatement. It extracts data from `parsed` result and save them as DatabaseSlowStatement. SlowSql will not abort or edit logs, you can use other LAL for further processing.
SlowSql will reuse `service`, `layer` and `timestamp` of extractor, so it is necessary to use `SlowSQL` after setting these.

> We need to set a log tag `"LOG_KIND" = "SLOW_SQL"` so that OAP can distinguish slow sql logs.

Reviewer 2: ```suggestion
We require a log tag `"LOG_KIND" = "SLOW_SQL"` to make OAP distinguish slow SQL logs from other log reports.
```

> A example of JSON sent to OAP is following:

Reviewer 2: ```suggestion
An example of JSON sent to OAP is as following:
```

``` json
[
   {
      "tags":{
         "data":[
            {
               "key":"LOG_KIND",
               "value":"SLOW_SQL"
            }
         ]
      },
      "body":{
         "json":{
            "json":"{\"time\":\"20220906153959\",\"id\":\"cb92c1a5b-2691e-fb2f-457a-9c72a392d9ed\",\"service\":\"root[root]@[localhost]\",\"statement\":\"select sleep(2);\",\"layer\":\"MYSQL\",\"query_time\":2000}"
         }
      },
      "service":"root[root]@[localhost]"
   }
]
```

- `statement`

> `statement` extracts the time bucket from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be

Reviewer: ```suggestion
`statement` extracts the SQL statement from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be
```

persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `latency`

`latency` extracts the latency from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be
persisted (if not dropped) and is used to associate with TopNDatabaseStatement.

- `id`

`id` extracts the id from the `parsed` result, and set it into the `DatabaseSlowStatement`, which will be persisted (if not
dropped) and is used to associate with TopNDatabaseStatement.

A Example of LAL to distinguish slow logs:

```groovy

> filter {
>         json{
>         }
>         extractor{
>           layer parsed.layer as String
>           service parsed.service as String
>           timestamp parsed.time as String
>           if (tag("LOG_KIND") == "SLOW_SQL") {
>              slowSql {
>                       id parsed.id as String
>                       statement parsed.statement as String
>                       latency parsed.query_time as Long
>                      }
>           }
>         }
> }

Reviewer: Let's format these lines

```suggestion
filter {
  json{
  }
  extractor{
    layer parsed.layer as String
    service parsed.service as String
    timestamp parsed.time as String
    if (tag("LOG_KIND") == "SLOW_SQL") {
      slowSql {
        id parsed.id as String
        statement parsed.statement as String
        latency parsed.query_time as Long
      }
    }
  }
}
```

    exp: http_response_time.sum(['le', 'service', 'instance']).increase('PT5M').histogram().histogram_percentile([50,70,90,99])
```

> - SlowSql

Reviewer: Just missed this, let’s keep the same style as others. 

```suggestion
- `slowSql`
```

> SlowSql aims to convert LogData to DatabaseSlowStatement. It extracts data from `parsed` result and save them as DatabaseSlowStatement. SlowSql will not abort or edit logs, you can use other LAL for further processing.

Reviewer: ```suggestion
`slowSql` aims to convert LogData to DatabaseSlowStatement. It extracts data from `parsed` result and save them as DatabaseSlowStatement. SlowSql will not abort or edit logs, you can use other LAL for further processing.
```
